### ISSUE TITLE
Implementing new family

### ISSUE THREAD

**Issue by BERENZ** [glmmTMB/glmmTMB#846]
I am new to `glmmTMB` and I really like what this package offers. The research problem I encounter requires implementing new family, i.e. zero-one truncated Poisson distribution (we observe only y>=2 counts) . Is there any reference that describe how to implement custom family or maybe the process of adding new family is complicated?

**Comment by bbolker** [comment]
It's moderately but not horribly complicated. I've been meaning to write up an example/vignette about this. I would say the main reason we haven't done anything like what you're suggesting is that if we were going to do that, we would like to come up with a more general form for truncation/censoring that could be used with user-specified truncation and censoring points, and with different distributions, which would be more work ...

**Comment by bbolker** [comment]
There is now a working (I think) example at https://github.com/glmmTMB/glmmTMB/tree/zotruncpois .  There's a vignette there too, but we have to think a little bit more about how to cherry-pick that into the main branch in such a way that it will run without the changes actually being made (i.e., set `eval=FALSE` for the example chunk and hard-code the output ...)

**Comment by BERENZ** [comment]
Thank you for the example and vignette! It is informative and very helpful. I'll check it out this week. In the meantime I started implementing these families in the [GLMMadaptive](https://github.com/drizopoulos/GLMMadaptive/) package.

**Comment by bbolker** [comment]
vignette added [here](https://github.com/glmmTMB/glmmTMB/pull/869). Will close once PR is merged.

**Comment by haticeakdur** [comment]
Dear Dr. Bolker,

I hope you are doing well.

For a new count distribution, I would like to develop a new mixed model.
The GLMMadaptive package allows defining mixed models for new distributions. I've carried a simulation study for a new count distribution, "Bell distribution", by defining the necessary functions of the distribution in the glmmAdaptive package. MSE and bias values were slightly higher than those of Poisson and negative binomal mixed models. Another problem is that I wanted to compare 3 models with AIC and log-lik. values on a real data set. For Bell mixed model: Fit statistics:
  log.Lik AIC BIC
  2388.24 -4768.481 -4763.768
; for Poisson: Fit statistics:
    log.Lik AIC BIC
  -650.9859 1309.972 1314.684
; for Neg Bin MM:
Fit statistics:
    log.Lik AIC BIC
  -499.0971 1008.194 1014.084
Bell MM  fit statistics values were quite different those in Poisson MM and Neg. Bin. MM. Do we have a chance to create this model in the glmmTMB package?

I've attached the bell distribution paper.
[On the Bell distribution and its associated regression model for count data.pdf](https://github.com/glmmTMB/glmmTMB/files/15339583/On.the.Bell.distribution.and.its.associated.regression.model.for.count.data.pdf)

Very happy if you help.

**Comment by bbolker** [comment]
Can you start by reading the [appropriate vignette](https://cran.r-project.org/web/packages/glmmTMB/vignettes/hacking.html) ?  Based on a quick glance at the paper, it looks like this shouldn't be too hard.   Unlike the [procedure for GLMMadaptive](https://drizopoulos.github.io/GLMMadaptive/articles/Custom_Models.html), you do have to code the log-likelihood for the new family in C++, but for simple expressions the code will look almost the way it would in R ...

**Comment by bbolker** [comment]
It looks like the hardest part of this will be to compute the Bell number, but there are a lot of places you can steal code from, e.g. here: https://www.geeksforgeeks.org/bell-numbers-number-of-ways-to-partition-a-set/ (for efficiency, it may be worth "memoizing", i.e. keeping a lookup table of values.  It would be better to find a stable way to compute the *log* of the Bell number, as the term `log(B(y)/y!` will overflow quickly (what is a typical maximum response value in your applications?

**Comment by haticeakdur** [comment]
I will be very happy if I manage to place these in the right place in the glmmTMB.cpp file... I haven't coded in C++ for a long time. I don't have experience combining R and C++, but I'll try... Thanks a lot for guiding me.

**Comment by bbolker** [comment]
The best way to do this will be to create your own forked repo on Github. When you've made as much progress as you can, you can push your results to your repo and ask us to have a look (either informally, or by submitting a pull request).

**Comment by haticeakdur** [comment]
It is a good idea. I'll let you know when I make progress and make ready the functions for bell family. Fingers crossed! Thank you!

**Comment by haticeakdur** [comment]
Dear Dr. Bolker
I created a fork on Github to work on glmmTMB with the bell distribution. I wrote the necessary functions in C++. I added the necessary codes just below poisson in glmmTMB. I tested the codes I wrote separately in C++ and they worked without any problems. I also added the family function for glmmTMB.r to implement it as a new function. I haven't combined it yet and made the package ready.  "https://github.com/haticeakdur/bell_TMB/tree/main"
I would appreciate it if you could take a look at the codes in the fork. If anything is missing or wrong, please let me know. Thank you in advance.

**Comment by bbolker** [comment]
Hi, this looks interesting, but I have a lot of questions/a lot of stuff doesn't look right. The main issues are:

* you need to generate your fork from the `glmmTMB/glmmTMB` repository, not from the generic place you cloned - that's the only way that a pull request will be feasible
* your code fails some basic tests (see my pull request to your repo)
* you need to implement your C++ code in the specific form required by glmmTMB ...

**Comment by haticeakdur** [comment]
Dear Dr. Bolker,

Thank you very much. I will work on that. I'm sorry I haven't worked on a github project with a team. I am not familiar terms of github coding platform. I'll learn and let you know on my progress.
Best!

**Comment by haticeakdur** [comment]
Dear Dr. Bolker
I've changed c++ functions of bell distribution to add new family. And I've tested them as you suggested. I add bell_test2 to github space and update the bell function file on github. May you add this family to glmmTMB package please for me. Because I am not capable of adding this family to the R package. Thank you in advance.
https://github.com/bbolker/bell_TMB/blob/main/bell_test2

**Comment by bbolker** [comment]
I'm going to open a new issue for this, since there are some interesting/challenging questions about how to implement it.

**Comment by ivangalvan** [comment]
Dear Dr. Bolker,

I am interested in using the zero-one truncated Poisson GLMM. I attempted to modify the glmmTMB package following your vignette, but encountered an error during the ```devtools:document()``` step:

```
   make: *** [/usr/local/lib/R/etc/Makeconf:204: glmmTMB.o] Error 1
   ERROR: compilation failed for package ‘glmmTMB’
─  removing ‘/tmp/RtmpZiPhIX/devtools_install_15660242fd3/glmmTMB’
Error in `(function (command = NULL, args = character(), error_on_status = TRUE, …`:
! System command 'R' failed
---
Exit status: 1
stdout & stderr: <printed>
---
Type .Last.error to see the more details.
```

Would you be able to provide a compiled and built version of the glmmTMB package with the zero-one truncated Poisson family included? This would allow me to download and install it on my machine.

Thank you for your time and assistance

**Comment by bbolker** [comment]
Hmm.  I suspect something is failing upstream, probably with your development tools (you do need to have compilers etc. installed).  What happens if you do `R CMD INSTALL <directory_name>` from the command line, where `<directory_name>` is the directory in which the package lives?

**Comment by ivangalvan** [comment]
Thank you for your quick reply. I get similar error, here the final lines:

```
/home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/tiny_ad/atomic.hpp:30:1:   required from ‘void atomic::bessel_kOp<order, ninput, noutput, mask>::reverse(TMBad::ReverseArgs<Type>&) [with Type = double; int order = 3; int ninput = 2; int noutput = 8; long int mask = 9]’
/home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/TMBad/global.hpp:1721:28:   required from ‘void TMBad::global::AddForwardMarkReverseMark<OperatorBase>::reverse(TMBad::ReverseArgs<Type>&) [with Type = double; OperatorBase = TMBad::global::AddIncrementDecrement<TMBad::global::AddDependencies<TMBad::global::AddInputSizeOutputSize<atomic::bessel_kOp<3, 2, 8, 9> > > >]’ /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/TMBad/global.hpp:2134:57:   required from ‘void TMBad::global::Complete<OperatorBase>::reverse(TMBad::ReverseArgs<double>&) [with OperatorBase = atomic::bessel_kOp<3, 2, 8, 9>]’                                                                                                                                               /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/TMBad/global.hpp:2134:10:   required from here   
/home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/DenseCoeffsBase.h:56:30: warning: ignoring attributes on template argument ‘Eigen::internal::packet_traits<double>::type’ {aka ‘__m128d’} [-Wignored-attributes] 
make: *** [/usr/lib/R/etc/Makeconf:200: glmmTMB.o] Error 1   
ERROR: compilation failed for package ‘glmmTMB’
* removing ‘/home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/glmmTMB’
* 
```

My gcc and g++ versions are: 

```gcc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0 ```
```g++ (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0 ```

**Comment by bbolker** [comment]
It may be very long, but can you post *all* of your output?

**Comment by ivangalvan** [comment]
I am sorry that for some reason, doing copy and paste the font size is huge (skipped many lines because indeed all output is long):

   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/Matrix.h:423:29: required from ‘Eigen::Matrix<_Scalar, _Rows, _Cols, _Options, _MaxRows, _MaxCols>::Matrix(const Eigen::EigenBase<OtherDerived>&) [with OtherDerived = Eigen::Product<Eigen::Matrix<double, 2, 4, 0, 2, 4>, Eigen::Matrix<double, 4, 1>, 0>; _Scalar = double; int _Rows = 2; int _Cols = 1; int _Options = 0; int _MaxRows = 2; int _MaxCols = 1]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/tiny_ad/atomic.hpp:82:1: required from ‘void atomic::logspace_subOp<order, ninput, noutput, mask>::reverse(TMBad::ReverseArgs<Type>&) [with Type = double; int order = 2; int ninput = 2; int noutput = 4; long int mask = 9]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/TMBad/global.hpp:1721:28: required from ‘void TMBad::global::AddForwardMarkReverseMark<OperatorBase>::reverse(TMBad::ReverseArgs<Type>&) [with Type = double; OperatorBase = TMBad::global::AddIncrementDecrement<TMBad::global::AddDependencies<TMBad::global::AddInputSizeOutputSize<atomic::logspace_subOp<2, 2, 4, 9> > > >]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/TMBad/global.hpp:2134:57: required from ‘void TMBad::global::Complete<OperatorBase>::reverse(TMBad::ReverseArgs<double>&) [with OperatorBase = atomic::logspace_subOp<2, 2, 4, 9>]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/TMBad/global.hpp:2134:10: required from here
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/DenseCoeffsBase.h:56:30:warning: noring attributes on template argument ‘Eigen::internal::packet_traits<double>::type{aka ‘__m128d-Wignored-attributes]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/DenseCoeffsBase.h: In instantiation of ‘class Eigen::DenseCoeffsBase<Eigen::CwiseBinaryOp<Eigen::internal::scalar_product_op<double, double>, const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, const Eigen::Matrix<double, -1, 1> >, const Eigen::MatrixWrapper<const Eigen::Array<double, -1, 1> > >, 0>
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/DenseBase.h:41:34: required from ‘class Eigen::DenseBase<Eigen::CwiseBinaryOp<Eigen::internal::scalar_product_op<double, double>, const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, const Eigen::Matrix<double, -1, 1> >, const Eigen::MatrixWrapper<const Eigen::Array<double, -1, 1> > > >
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/MatrixBase.h:48:34: required from ‘class Eigen::MatrixBase<Eigen::CwiseBinaryOp<Eigen::internal::scalar_product_op<double, double>, const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, const Eigen::Matrix<double, -1, 1> >, const Eigen::MatrixWrapper<const Eigen::Array<double, -1, 1> > > >
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/CwiseBinaryOp.h:148:7: required from ‘class Eigen::CwiseBinaryOpImpl<Eigen::internal::scalar_product_op<double, double>, const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, const Eigen::Matrix<double, -1, 1> >, const Eigen::MatrixWrapper<const Eigen::Array<double, -1, 1> >, Eigen::Dense>
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/CwiseBinaryOp.h:77:7: required from ‘class Eigen::CwiseBinaryOp<Eigen::internal::scalar_product_op<double, double>, const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, const Eigen::Matrix<double, -1, 1> >, const Eigen::MatrixWrapper<const Eigen::Array<double, -1, 1> > >
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/ProductEvaluators.h:280:48: required from ‘void Eigen::internal::outer_product_selector_run(Dst&, const Lhs&, const Rhs&, const Func&, const Eigen::internal::false_type&) [with Dst = Eigen::Matrix<double, -1, -1>; Lhs = Eigen::MatrixWrapper<const Eigen::Array<double, -1, 1> >; Rhs = Eigen::Transpose<const Eigen::MatrixWrapper<const Eigen::Array<double, -1, 1> > >; Func = Eigen::internal::generic_product_impl<Eigen::MatrixWrapper<const Eigen::Array<double, -1, 1> >, Eigen::Transpose<const Eigen::MatrixWrapper<const Eigen::Array<double, -1, 1> > >, Eigen::DenseShape, Eigen::DenseShape, 5>::set]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/ProductEvaluators.h:317:41: [ skipping 12 instantiation contexts, use -ftemplate-backtrace-limit=0 to disable ]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/tmbutils/newton.hpp:93:54: required from ‘newton::matrix<Type>::matrix(const Eigen::ArrayBase<OtherDerived>&) [with Derived = Eigen::ArrayWrapper<const Eigen::Product<Eigen::MatrixWrapper<const Eigen::Array<double, -1, 1> >, Eigen::Transpose<const Eigen::MatrixWrapper<const Eigen::Array<double, -1, 1> > >, 0> >; Type = double]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/tmbutils/newton.hpp:222:15: required from ‘newton::vector<T> newton::jacobian_dense_t<Factorization>::crossprod(const newton::vector<T>&, const newton::vector<T>&) [with T = double; Factorization = Eigen::LLT<Eigen::Matrix<double, -1, -1>, 1>]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/tmbutils/newton.hpp:163:45: required from ‘void newton::HessianSolveVector<Hessian_Type>::reverse(TMBad::ReverseArgs<Type>&) [with T = double; Hessian_Type = newton::jacobian_dense_t<>]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/TMBad/global.hpp:1721:28: required from ‘void TMBad::global::AddForwardMarkReverseMark<OperatorBase>::reverse(TMBad::ReverseArgs<Type>&) [with Type = double; OperatorBase = TMBad::global::AddIncrementDecrement<TMBad::global::AddDependencies<newton::HessianSolveVector<newton::jacobian_dense_t<> > > >]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/TMBad/global.hpp:2134:57: required from ‘void TMBad::global::Complete<OperatorBase>::reverse(TMBad::ReverseArgs<double>&) [with OperatorBase = newton::HessianSolveVector<newton::jacobian_dense_t<> >]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/TMBad/global.hpp:2134:10: required from here
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/DenseCoeffsBase.h:56:30:warning: noring attributes on template argument ‘Eigen::internal::packet_traits<double>::type{aka ‘__m128d-Wignored-attributes]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/DenseCoeffsBase.h: In instantiation of ‘class Eigen::DenseCoeffsBase<Eigen::Matrix<double, 2, 8, 0, 2, 8>, 0>
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/DenseCoeffsBase.h:302:7: required from ‘class Eigen::DenseCoeffsBase<Eigen::Matrix<double, 2, 8, 0, 2, 8>, 1>
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/DenseCoeffsBase.h:555:7: required from ‘class Eigen::DenseCoeffsBase<Eigen::Matrix<double, 2, 8, 0, 2, 8>, 3>
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/DenseBase.h:41:34: required from ‘class Eigen::DenseBase<Eigen::Matrix<double, 2, 8, 0, 2, 8> >
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/MatrixBase.h:48:34: required from ‘class Eigen::MatrixBase<Eigen::Matrix<double, 2, 8, 0, 2, 8> >
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/PlainObjectBase.h:98:7: required from ‘class Eigen::PlainObjectBase<Eigen::Matrix<double, 2, 8, 0, 2, 8> >
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/Matrix.h:178:7: required from ‘class Eigen::Matrix<double, 2, 8, 0, 2, 8>
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/tiny_ad/atomic.hpp:30:1: required from ‘void atomic::bessel_kOp<order, ninput, noutput, mask>::reverse(TMBad::ReverseArgs<Type>&) [with Type = double; int order = 3; int ninput = 2; int noutput = 8; long int mask = 9]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/TMBad/global.hpp:1721:28: required from ‘void TMBad::global::AddForwardMarkReverseMark<OperatorBase>::reverse(TMBad::ReverseArgs<Type>&) [with Type = double; OperatorBase = TMBad::global::AddIncrementDecrement<TMBad::global::AddDependencies<TMBad::global::AddInputSizeOutputSize<atomic::bessel_kOp<3, 2, 8, 9> > > >]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/TMBad/global.hpp:2134:57: required from ‘void TMBad::global::Complete<OperatorBase>::reverse(TMBad::ReverseArgs<double>&) [with OperatorBase = atomic::bessel_kOp<3, 2, 8, 9>]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/TMBad/global.hpp:2134:10: required from here
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/DenseCoeffsBase.h:56:30:warning: noring attributes on template argument ‘Eigen::internal::packet_traits<double>::type{aka ‘__m128d-Wignored-attributes]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/DenseCoeffsBase.h: In instantiation of ‘class Eigen::DenseCoeffsBase<Eigen::Product<Eigen::Matrix<double, 2, 8, 0, 2, 8>, Eigen::Matrix<double, 8, 1, 0, 8, 1>, 0>, 0>
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/DenseBase.h:41:34: required from ‘class Eigen::DenseBase<Eigen::Product<Eigen::Matrix<double, 2, 8, 0, 2, 8>, Eigen::Matrix<double, 8, 1, 0, 8, 1>, 0> >
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/MatrixBase.h:48:34: required from ‘class Eigen::MatrixBase<Eigen::Product<Eigen::Matrix<double, 2, 8, 0, 2, 8>, Eigen::Matrix<double, 8, 1, 0, 8, 1>, 0> >
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/Product.h:120:7: required from ‘class Eigen::internal::dense_product_base<Eigen::Matrix<double, 2, 8, 0, 2, 8>, Eigen::Matrix<double, 8, 1, 0, 8, 1>, 0, 3>
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/Product.h:152:7: required from ‘class Eigen::ProductImpl<Eigen::Matrix<double, 2, 8, 0, 2, 8>, Eigen::Matrix<double, 8, 1, 0, 8, 1>, 0, Eigen::Dense>
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/Product.h:71:7: required from ‘class Eigen::Product<Eigen::Matrix<double, 2, 8, 0, 2, 8>, Eigen::Matrix<double, 8, 1, 0, 8, 1>, 0>
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/tiny_ad/atomic.hpp:30:1: required from ‘void atomic::bessel_kOp<order, ninput, noutput, mask>::reverse(TMBad::ReverseArgs<Type>&) [with Type = double; int order = 3; int ninput = 2; int noutput = 8; long int mask = 9]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/TMBad/global.hpp:1721:28: required from ‘void TMBad::global::AddForwardMarkReverseMark<OperatorBase>::reverse(TMBad::ReverseArgs<Type>&) [with Type = double; OperatorBase = TMBad::global::AddIncrementDecrement<TMBad::global::AddDependencies<TMBad::global::AddInputSizeOutputSize<atomic::bessel_kOp<3, 2, 8, 9> > > >]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/TMBad/global.hpp:2134:57: required from ‘void TMBad::global::Complete<OperatorBase>::reverse(TMBad::ReverseArgs<double>&) [with OperatorBase = atomic::bessel_kOp<3, 2, 8, 9>]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/TMBad/global.hpp:2134:10: required from here
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/DenseCoeffsBase.h:56:30:warning: noring attributes on template argument ‘Eigen::internal::packet_traits<double>::type{aka ‘__m128d-Wignored-attributes]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/DenseCoeffsBase.h: In instantiation of ‘class Eigen::DenseCoeffsBase<Eigen::Product<Eigen::Matrix<double, 2, 8, 0, 2, 8>, Eigen::Matrix<double, 8, 1, 0, 8, 1>, 1>, 0>
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/DenseBase.h:41:34: required from ‘class Eigen::DenseBase<Eigen::Product<Eigen::Matrix<double, 2, 8, 0, 2, 8>, Eigen::Matrix<double, 8, 1, 0, 8, 1>, 1> >
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/MatrixBase.h:48:34: required from ‘class Eigen::MatrixBase<Eigen::Product<Eigen::Matrix<double, 2, 8, 0, 2, 8>, Eigen::Matrix<double, 8, 1, 0, 8, 1>, 1> >
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/Product.h:120:7: required from ‘class Eigen::internal::dense_product_base<Eigen::Matrix<double, 2, 8, 0, 2, 8>, Eigen::Matrix<double, 8, 1, 0, 8, 1>, 1, 3>
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/Product.h:152:7: required from ‘class Eigen::ProductImpl<Eigen::Matrix<double, 2, 8, 0, 2, 8>, Eigen::Matrix<double, 8, 1, 0, 8, 1>, 1, Eigen::Dense>
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/Product.h:71:7: required from ‘class Eigen::Product<Eigen::Matrix<double, 2, 8, 0, 2, 8>, Eigen::Matrix<double, 8, 1, 0, 8, 1>, 1>
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/ProductEvaluators.h:402:50: [ skipping 4 instantiation contexts, use -ftemplate-backtrace-limit=0 to disable ]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/PlainObjectBase.h:594:19: required from ‘Eigen::PlainObjectBase<Derived>::PlainObjectBase(const Eigen::DenseBase<OtherDerived>&) [with OtherDerived = Eigen::Product<Eigen::Matrix<double, 2, 8, 0, 2, 8>, Eigen::Matrix<double, 8, 1, 0, 8, 1>, 0>; Derived = Eigen::Matrix<double, 2, 1>]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/Matrix.h:423:29: required from ‘Eigen::Matrix<_Scalar, _Rows, _Cols, _Options, _MaxRows, _MaxCols>::Matrix(const Eigen::EigenBase<OtherDerived>&) [with OtherDerived = Eigen::Product<Eigen::Matrix<double, 2, 8, 0, 2, 8>, Eigen::Matrix<double, 8, 1, 0, 8, 1>, 0>; _Scalar = double; int _Rows = 2; int _Cols = 1; int _Options = 0; int _MaxRows = 2; int _MaxCols = 1]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/tiny_ad/atomic.hpp:30:1: required from ‘void atomic::bessel_kOp<order, ninput, noutput, mask>::reverse(TMBad::ReverseArgs<Type>&) [with Type = double; int order = 3; int ninput = 2; int noutput = 8; long int mask = 9]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/TMBad/global.hpp:1721:28: required from ‘void TMBad::global::AddForwardMarkReverseMark<OperatorBase>::reverse(TMBad::ReverseArgs<Type>&) [with Type = double; OperatorBase = TMBad::global::AddIncrementDecrement<TMBad::global::AddDependencies<TMBad::global::AddInputSizeOutputSize<atomic::bessel_kOp<3, 2, 8, 9> > > >]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/TMBad/global.hpp:2134:57: required from ‘void TMBad::global::Complete<OperatorBase>::reverse(TMBad::ReverseArgs<double>&) [with OperatorBase = atomic::bessel_kOp<3, 2, 8, 9>]
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/TMB/include/TMBad/global.hpp:2134:10: required from here
   /home/igalvan/R/x86_64-pc-linux-gnu-library/4.3/RcppEigen/include/Eigen/src/Core/DenseCoeffsBase.h:56:30:warning: noring attributes on template argument ‘Eigen::internal::packet_traits<double>::type{aka ‘__m128d-Wignored-attributes]
   make: *** [/usr/lib/R/etc/Makeconf:200: glmmTMB.o] Error 1
   ERROR: compilation failed for package ‘glmmTMB’
─  removing ‘/home/igalvan/Documents/tmp_R/RtmpmiYHcx/devtools_install_203ff1999ffc0/glmmTMB’
Error in `(function (command = NULL, args = character(), error_on_status = TRUE, …`:
! System command 'R' failed
---
Exit status: 1
stdout & stderr: <printed>
---
Type .Last.error to see the more details.

**Comment by bbolker** [comment]
Can you try re-installing RcppEigen and TMB, ideally from source? (I note that you're on the last major release (4.3.x; 4.4.0 was released in April 2024), which *might* cause some of the problems ...)

**Comment by ivangalvan** [comment]
I have the same error after re-installing RcppEigen_0.3.4.0.2 and TMB_1.9.15 from source. I found curious that the devtools::document() is working fine if I do not include the case zo_truncated_poisson_family. Maybe I am not copying and pasting the zo_truncated_poisson_family case into the glmmTMB.cpp correctly? I included it just bellow the bell case:

```
      case bell_family:
	// unfortunately need to back-transform from mu to underlying theta via Lambert W ...

	// see https://stackoverflow.com/questions/92396/why-cant-variables-be-declared-in-a-switch-statement for {} 
	{
	  Type btheta;
	  btheta = glmmtmb::LambertW(mu(i));
	  tmp_loglik = glmmtmb::dbell(yobs(i), btheta, true);
	  SIMULATE{yobs(i) = glmmtmb::rbell(btheta);}
	  break;
	}
      case zo_truncated_poisson_family:
        log_nzprob = logspace_sub(Type(0), -mu(i));  // log(1-exp(-mu(i)));
        // now subtract the prob(X==1)
        log_nzprob = logspace_sub(log_nzprob, log(mu(i)) - mu(i));
        // log-Poisson likelihood minus the 'missing mass'
        tmp_loglik = dpois(yobs(i), mu(i), true) - log_nzprob;
        // this is a utility for use in ther zero-inflated case
        tmp_loglik = zt_lik_nearzero(yobs(i), tmp_loglik);
        SIMULATE{
          // conveniently, this built-in function already allows truncation
          //  at different points
          yobs(i) = glmmtmb::rtruncated_poisson(1, asDouble(mu(i)));
        }
      break;        
      default:
        error("Family not implemented!");
      } // End switch
```

**Comment by bbolker** [comment]
`log_nzprob` is defined as a vector, not a scalar ... when I replaced all instances of `log_nzprob` in your code chunk with `log_nzprob(i)`, I was able to compile/install successfully (but haven't done any further testing). (The instructions should probably say that you should make sure your C++ code compiles before proceeding to the next step ...)

For what it's worth, I figured this out by looking at the **first** error in the giant stream of error output:

```
glmmTMB.cpp: In instantiation of ‘Type objective_function<Type>::operator()() [with Type = TMBad::global::ad_aug]’:
/usr/lib/R/site-library/TMB/include/tmb_core.hpp:1277:6:   required from here
glmmTMB.cpp:1049:34: error: no matching function for call to ‘logspace_sub(tmbutils::vector<TMBad::global::ad_aug>&, TMBad::ad_aug)’
 1049 |         log_nzprob = logspace_sub(log_nzprob, log(mu(i)) - mu(i));
      |                      ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by ivangalvan** [comment]
Great. Thank you. When replycing ```log_nzprob``` by ```log_nzprob(i)``` I was able to run ```devtools::document()``` properly. Then, I made ```make enum-update``` by adding in my working directory the Makefile downloaded from the Github repository. And finally, I installed the package: ```install.packages("glmmTMB",res=NULL) ```. However, I am not able to run the example code:

````
library(glmmTMB)
set.seed(101)
dd <- data.frame(y = rpois(500, exp(1)))
table(dd$y)
##  0   1   2   3   4   5   6   7   8   9 
## 34  91 117 116  68  45  17   7   3   2 
dd <- dd[dd$y>1,,drop=FALSE]
table(dd$y)
##   2   3   4   5   6   7   8   9 
## 117 116  68  45  17   7   3   2 
glmmTMB(y ~ 1, family = "zo_truncated_poisson", data = dd)
````

Error in get(family, mode = "function", envir = parent.frame(2)) :                                                                                                                              object 'zo_truncated_poisson' of mode 'function' was not found

**Comment by bbolker** [comment]
Did you implement the machinery on the R side? (I think this is discussed in the vignette?)

**Comment by ivangalvan** [comment]
I re-installed by typing ```R CMD INSTALL glmmTMB``` and now the provided example is working. Thank you for your help.

**Comment by ivangalvan** [comment]
I realized that I am obtaining fitted values from zero to two. There is something we miss in the truncation of the glmmTMB.cpp? I also included the variance in the family.R:

```
#' @rdname nbinom2
#' @export
zo_truncated_poisson <- function(link="log") {
  r <- list(family="zo_truncated_poisson",
            variance=function(lambda) {
              stop("haven't implemented variance function")
              ## should figure this out ...
              (lambda - lambda * exp(-lambda)) / (1 - exp(-lambda) - lambda * exp(-lambda))
            })
  return(make_family(r,link))
}
```
